src/markdown_exec/formatters/jupyter.py

```diff
@@ -0,0 +1,253 @@
+"""Formatter for executing code using Jupyter kernels."""
+
+from __future__ import annotations
+
+import atexit
+import functools
+import os
+import platform
+import queue
+import time
+from typing import TYPE_CHECKING, Any, Callable
+
+from jupyter_client import KernelManager
+from jupyter_client.kernelspec import KernelSpecManager
+
+from markdown_exec.formatters.base import ExecutionError, base_format
+from markdown_exec.rendering import code_block
+
+if TYPE_CHECKING:
+    from jupyter_client.blocking.client import BlockingKernelClient
+
+_SHUTDOWN_TIMEOUT = 10
+
+# Store kernel managers by session and kernel name
+_kernel_managers: dict[str, dict[tuple[str | None, str], KernelManager]] = {}
+_kernel_clients: dict[str, dict[tuple[str | None, str], BlockingKernelClient]] = {}
+
+
+def _get_available_kernels() -> dict[str, dict[str, Any]]:
+    """Get information about all available Jupyter kernels.
+
+    Returns:
+        A dictionary mapping kernel names to their information including:
+        - display_name: Human readable name
+        - language: The programming language
+        - language_version: Version of the language (if available)
+        - language_info: Additional language information
+    """
+    ksm = KernelSpecManager()
+    # Find all available kernel specs
+    kernel_specs = ksm.get_all_specs()
+
+    # Get detailed information for each kernel
+    return {
+        name: {
+            "display_name": spec["spec"]["display_name"],
+            "languages": [
+                spec["spec"]["language"],
+                *(spec["spec"].get("aliases", [])),
+            ],
+            "language_version": spec["spec"].get("language_version", None),
+            "language_info": spec["spec"].get("language_info", {}),
+        }
+        for name, spec in kernel_specs.items()
+    }
+
+_language_to_kernel: dict[str, str] = {}
+
+def _get_kernel_for_language(language: str) -> str:
```

Comment:
style: Looks like this code is not Ruff-formatted, could you run `make format` :smile:?